src/crc.rs

```diff
@@ -0,0 +1,56 @@
+use core::fmt;
+use std::error::Error;
+
+// CRC table generation and Mode S checksumming ported from
+// https://github.com/wiedehopf/readsb/blob/177545fbff8cc2be9d7e9b9b109c5c1046c2642b/crc.c
+
+const MODES_GENERATOR_POLY: u32 = 0xfff409;
+
+lazy_static! {
+    static ref CRC_TABLE: [u32; 256] = {
+        let mut result = [0u32; 256];
+        for i in 0..256 {
+            let mut c = i << 16;
+            for _j in 0..8 {
+                if c & 0x800000 != 0 {
+                    c = (c << 1) ^ MODES_GENERATOR_POLY;
+                } else {
+                    c <<= 1;
+                }
+            }
+            result[i as usize] = c & 0x00ffffff;
+        }
+        result
+    };
+}
+
+#[derive(Debug, Clone)]
+pub struct CrcError(String);
+
+impl fmt::Display for CrcError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl Error for CrcError {}
+
+pub fn mode_s_crc(input: &[u8], num_bytes: u8) -> Result<u32, CrcError> {
+    let mut rem = 0u32;
+    if num_bytes < 3 {
+        return Err(CrcError(String::from(
```

Comment:
Please see my other comment where this is called. Given the lack of custom error handling in the parsers I would replace the `CrcError` type with a simple `Option` for now. I want to improve the surfacing of custom parsing errors but that shouldn't block merging this.